### PR TITLE
Setting KEDA specific annotations from KafkaChannel

### PR DESCRIPTION
**Pull request by aavarghese**
Signed-off-by: aavarghese <avarghese@us.ibm.com>

## Proposed Changes

- Setting KEDA specific annotations from `KafkaChannel` into ConsumerGroup API
- **If KEDA is installed**, `ConsumerGroup` replicas field is set by KEDA controller (no initial value)

### CODE REVIEW

**Comment by aavarghese** [comment]
Waiting for https://github.com/knative-sandbox/eventing-kafka-broker/pull/2607 and https://github.com/knative-sandbox/eventing-kafka-broker/pull/2608 to merge first.

Part of breaking down https://github.com/knative-sandbox/eventing-kafka-broker/pull/2437 into smaller ones.
This third one is ready for review and merge.

**Comment by aavarghese** [comment]
```
2022-09-13T08:54:07Z	ERROR	Target resource doesn't expose /scale subresource	{"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "scaledObject": {"name":"so-b00640a1-b65e-445d-926d-52787713e0d6","namespace":"default"}, "namespace": "default", "name": "so-b00640a1-b65e-445d-926d-52787713e0d6", "reconcileID": "77b37b08-1460-405d-9aad-fe050c526cf1", "resource": "internal.kafka.eventing.knative.dev/v1alpha1.ConsumerGroup", "name": "5e4e9b73-927d-4a5a-a7b3-1f3c8d9ac0da", "error": "Internal error occurred: the spec replicas field \".spec.replicas\" does not exist"}
github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).reconcileScaledObject
	/workspace/controllers/keda/scaledobject_controller.go:224
github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).Reconcile
	/workspace/controllers/keda/scaledobject_controller.go:186
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234
2022-09-13T08:54:07Z	ERROR	ScaledObject doesn't have correct scaleTargetRef specification	{"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "scaledObject": {"name":"so-b00640a1-b65e-445d-926d-52787713e0d6","namespace":"default"}, "namespace": "default", "name": "so-b00640a1-b65e-445d-926d-52787713e0d6", "reconcileID": "77b37b08-1460-405d-9aad-fe050c526cf1", "error": "Internal error occurred: the spec replicas field \".spec.replicas\" does not exist"}
```

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (pierDipi) [inline control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: should we propagate subscription annotations more than channel ones to have more granularity and consistently like triggers?

Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: Hmm is there an advantage to that granularity? Do we expect subscribers for the same channel to have different KEDA settings?

Comment (pierDipi) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: Yes, because for example different consumer want to move at different pace based on how fast they can process events, so they might want to set min and max replicas differently

Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: Ok. Though a channel type does not have the full `Subscribable` resource though (if its even called a resource) just the addressable info from the Spec. I assume it can be of any K8 type so getting the actual object to find its annotations might not be something we can do?

**Comment by aavarghese** [comment]
Channel dispatcher Java code throwing this error:
```
{"@timestamp":"2022-09-14T22:24:07.914Z","@version":"1","message":"Failed to startup the dispatcher","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.main.Main","thread_name":"main","level":"ERROR","level_value":40000,"stack_trace":"java.nio.file.NoSuchFileException: /etc/channels-subscriptions\n\tat java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)\n\tat java.base/sun.nio.fs.UnixException.asIOException(UnixException.java:115)\n\tat java.base/sun.nio.fs.LinuxWatchService$Poller.implRegister(LinuxWatchService.java:246)\n\tat java.base/sun.nio.fs.AbstractPoller.processRequests(AbstractPoller.java:266)\n\tat java.base/sun.nio.fs.LinuxWatchService$Poller.run(LinuxWatchService.java:364)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\n"}

```

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: We can, and that's what I'm doing for https://github.com/knative-sandbox/eventing-kafka-broker/pull/2651

Comment (pierDipi) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: I think we can extract a common function

Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:587]: My example KafkaChannel (that works) looks like this:
```
apiVersion: messaging.knative.dev/v1beta1
kind: KafkaChannel
metadata:
  name: kafka-channel
  annotations:
    autoscaling.knative.dev/class: keda.autoscaling.knative.dev
    autoscaling.knative.dev/minScale: "0" 
    autoscaling.knative.dev/maxScale: "5" 
    keda.autoscaling.knative.dev/pollingInterval: "3" 
    keda.autoscaling.knative.dev/cooldownPeriod: "10" 
    keda.autoscaling.knative.dev/kafkaLagThreshold: "2"  
spec:
  numPartitions: 10
  replicationFactor: 3
  
  :
    - subscriberUri: http://eventdisplay.default.svc.cluster.local/
      uid: "5e4e9b73-927d-4a5a-a7b3-1f3c8d9ac0da"
      
kind: Service
apiVersion: v1
metadata:
  name: eventdisplay
  annotations:
      prometheus.io/scrape: 'true'
      prometheus.io/port:   '8001'
spec:
  selector:
    app: eventdisplay
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8080
      name: http
```
where `subscribers.ui` is the UID of the K8 service, and not a `Subscription` object

```
 ansuvarghese$ k get subscriptions --all-namespaces
No resources found
```

Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:614]: @pierDipi wdyt about this kinda channel installs?

Comment (pierDipi) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go]: I'd say that when a subscriber is created without a subscription it cannot be scaled with KEDA

I don't even know if that should be even valid.

**Comment by aavarghese** [comment]
@pierDipi @matzew this channelv2 specific PR is ready for a review. KEDA related scaling objects/trigger auth changes for channels are in separate PRs...

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (pierDipi) [inline control-plane/pkg/reconciler/channel/v2/channelv2.go:610]: Can we remove this commented line?

File: data-plane/config/channelv2/500-dispatcher.yaml
Comment (pierDipi) [inline data-plane/config/channelv2/500-dispatcher.yaml:26]: Is it temporary? perhaps can be removed before merging?

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (pierDipi) [inline control-plane/pkg/reconciler/channel/v2/channelv2.go:599]: I think this will end up overriding whatever KEDA has set because 
```
if equality.Semantic.DeepDerivative(expectedCg.Spec, cg.Spec) ...
```
will be always false.

If it's not done yet, I think this should be done at the webhook level if replicas is not set.

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:610]: Yes, its not there anymore in the latest code..

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:560]: Also not there in the latest commit 8ee0ae8ed9542212f13af2fb5f7192ed150058bb

File: data-plane/config/channelv2/500-dispatcher.yaml
Comment (aavarghese) [reply data-plane/config/channelv2/500-dispatcher.yaml:165]: Because I found `DATA_PLANE_CONFIG_FILE_PATH` is at `/etc/channels-subscriptions/data` which is mounted at `kafka-channel-channels-subscriptions` .... @pierDipi

File: data-plane/config/channelv2/500-dispatcher.yaml
Comment (aavarghese) [reply data-plane/config/channelv2/500-dispatcher.yaml:26]: Wdyt about initializing dispatcher pods with 0 replicas so that we leave it to the scheduler's autoscaler to decide on the initial scaling value? Otherwise the [default is 1](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#replicas) and most of the times the installation has to wait till the dispatcher pod terminates when there's no kafkasources installed....

Comment (aavarghese) [reply data-plane/config/channelv2/500-dispatcher.yaml:26]: @pierDipi

File: control-plane/pkg/reconciler/channel/v2/channelv2.go
Comment (aavarghese) [reply control-plane/pkg/reconciler/channel/v2/channelv2.go:599]: Don't think its done at the webhook level today. Good pt

**Comment by aavarghese** [comment]
@pierDipi this one's past the flaky tests and now ready..

**Comment by aavarghese** [comment]
@pierDipi this PR is ready for review. Secrets both legacy and new work with KEDA integration for channels.

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
